jwt/keyset.go

```diff
@@ -0,0 +1,217 @@
+package jwt
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"net/http"
+
+	"github.com/coreos/go-oidc"
+	"github.com/hashicorp/go-cleanhttp"
+	"golang.org/x/oauth2"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+// KeySet represents a set of keys that can be used to verify the signatures of JWTs.
+// A KeySet is expected to be backed by a set of local or remote keys.
+type KeySet interface {
+
+	// VerifySignature parses the given JWT, verifies its signature, and returns the claims in its payload.
+	VerifySignature(ctx context.Context, token string) (claims map[string]interface{}, err error)
+}
+
+// OIDCDiscoveryKeySet verifies JWT signatures using keys obtained by the OIDC discovery mechanism.
+type OIDCDiscoveryKeySet struct {
+	provider *oidc.Provider
+}
+
+// JSONWebKeySet verifies JWT signatures using keys obtained from a JWKS URL.
+type JSONWebKeySet struct {
+	remoteJWKS oidc.KeySet
+}
+
+// StaticKeySet verifies JWT signatures using local PEM-encoded public keys.
+type StaticKeySet struct {
+	publicKeys []interface{}
+}
+
+// NewOIDCDiscoveryKeySet returns a KeySet that verifies JWT signatures using keys from the
+// JSON Web Key Set (JWKS) published in the discovery document at the given discoveryURL.
+// The client used to obtain the remote keys will verify server certificates using the root
+// certificates provided by discoveryCAPEM.
+func NewOIDCDiscoveryKeySet(ctx context.Context, discoveryURL string, discoveryCAPEM string) (KeySet, error) {
+	if discoveryURL == "" {
+		return nil, errors.New("discoveryURL must not be empty")
+	}
+
+	caCtx, err := createCAContext(ctx, discoveryCAPEM)
+	if err != nil {
+		return nil, err
+	}
+
+	provider, err := oidc.NewProvider(caCtx, discoveryURL)
+	if err != nil {
+		return nil, err
+	}
+
+	return &OIDCDiscoveryKeySet{
+		provider: provider,
+	}, nil
+}
+
+// VerifySignature parses the given JWT, verifies its signature using discovered JWKS keys, and
+// returns the claims in its payload. The given JWT must be of the JWS compact serialization form.
+func (ks OIDCDiscoveryKeySet) VerifySignature(ctx context.Context, token string) (map[string]interface{}, error) {
```

Comment:
I wonder if this is needed... could we just do something like: https://github.com/hashicorp/cap/tree/jimlambrt-jwt-idea

It would leverage the JSONWebKeySet type you have to have... and I don't see any need for the other capabilities provided by the oidc.Provider

Reply:
I like this idea. I don't think we need to hold on to `oidc.Provider` when all it really does is retrieve the `jwks_uri` from the OIDC discovery document and create a remote key set for verification. This would also make the verification more consistent with `JSONWebKeySet`. 

Thanks for putting together the example code!

Reply:
I've removed the need for using an `oidc.Provider` for obtaining keys via OIDC discovery in abb14cd. The code is very similar to that in your example branch with some minor adjustments and comments. Let me know your thoughts 👍